compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3689,7 +3689,7 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
                     }
                 }
                 // readonly can match to a union similar to any|error
-                if (sMember.tag == TypeTags.READONLY) {
+                if (sMember.tag == TypeTags.READONLY && isAssignable(symTable.anyAndReadonlyOrError, targetUnion)) {
                     unresolvedTypes.add(new TypePair(sMember, targetUnion));
                     if (isAssignable(sMember, targetUnion, unresolvedTypes)) {
```

Comment:
We can remove this second `isAssignable ` check if condition. Keep the logic inside the if block.

Reply:
Fixed.